### ISSUE TITLE
Vector and LA::distribued::Vector do not accept integral types

### ISSUE THREAD

**Issue by marcfehling** [dealii/dealii#11875]
We currently can not instantiate `Vector` and `LA::distributed::Vector` for integral types.

The issue lies in calculating norms. We use functions from the `std` library that do not work with these types, i.e.,
- `std::abs` does not work with unsigned types
- `std::sqrt` returns a double and complains if it gets implicitly converted into an integral type
- and probably more

PR #11857 relies on that feature, but currently falls back on using `float` as a hotfix.

**Comment by drwells** [comment]
We discussed this recently and I think that using the `Partitioner` classes directly is the right fix (uniform or nonuniform). If you want to move cell data around then this works pretty well, though the interface might be a bit clunky. Do you actually need to do any linear algebra? The only usage of this I saw in #11857 was for moving future levels.

**Comment by marcfehling** [comment]
> Do you actually need to do any linear algebra?

No, not for this purpose. It is just very easy to use the `distributed::Vector` class and exchange data. I'll figure out how to use the `Partitioner` classes.

**Comment by marcfehling** [comment]
I'll get some inspiration from [tests/mpi/parallel_partitioner_06](https://github.com/dealii/dealii/blob/master/tests/mpi/parallel_partitioner_06.cc).

**Comment by drwells** [comment]
I also have some code that moves around integers associated with cells (but with a nonuniform partitioner - the API is the same) here:

https://github.com/drwells/fiddle/blob/master/source/interaction/interaction_base.cc#L301

**Comment by kronbichler** [comment]
I agree with @drwells, using things directly via the partitioner class is the right approach. I realize that the interface provided by the partitioner requests one to manually keep track of the arrays and thus some extra work. If we feel for it, we might create a data container similar to `AlignedVector` (or `std::vector`) that can handle individual member access and data exchange according to partitioners, but without the vector space attribute of norms and other arithmetic operations. (That class could then even be a base class for `LA:d:V` if we want to avoid code duplication.)

**Comment by bangerth** [comment]
What you're really looking for is a "distributed array" based on a partitioner. The `LA::d::V` class is then really just such a distributed array that adds linear algebra functionality to it -- so as mentioned above could be a derived class. (I'm not advocating that @marcfehling actually write such a distributed array, just clarifying the notion of what @kronbichler already said.)

**Comment by marcfehling** [comment]
> If we feel for it, we might create a data container similar to `AlignedVector` (or `std::vector`) that can handle individual member access and data exchange according to partitioners, but without the vector space attribute of norms and other arithmetic operations. (That class could then even be a base class for `LA:d:V` if we want to avoid code duplication.)

> What you're really looking for is a "distributed array" based on a partitioner.

That would certainly be a nice feature to have! I can imagine that a Container type like this would come in handy in many places in the library, i.e., for storing `active_fe_indices` in `DoFHandler`.

**Comment by peterrum** [comment]
This seems to be a recurring topic, see also https://github.com/dealii/dealii/pull/12312#issuecomment-850474932.

**Comment by drwells** [comment]
Are other people still interested in this? I'd like to use exactly this functionality with global active cell index partitioner to simplify sharing some non-scalar cell data.

If someone wrote a draft I'd be happy to take it over.
